File: py/2021/day3/part2.py

from AOC import read_input

inp = read_input()
inp = inp.splitlines()

bit_pos = list(zip(*inp))

bit_crit_oxygene = inp[:]
b = list(zip(*bit_crit_oxygene))
for i in range(len(bit_pos)):
    if b[i].count('1') >= b[i].count('0'):
        c = '1'
    else:
        c = '0'
    bit_crit_oxygene = list(filter(lambda x: x[i] == c, bit_crit_oxygene))
    if len(bit_crit_oxygene) == 1:
        break
    b = list(zip(*bit_crit_oxygene))

oxygene = bit_crit_oxygene[0]

bit_crit_co2 = inp[:]
b = list(zip(*bit_crit_co2))
for i in range(len(bit_pos)):
    if b[i].count('1') < b[i].count('0'):
        c = '1'
    else:
        c = '0'
    bit_crit_co2 = list(filter(lambda x: x[i] == c, bit_crit_co2))
    if len(bit_crit_co2) == 1:
        break
    b = list(zip(*bit_crit_co2))

co2 = bit_crit_co2[0]

print(int(oxygene, 2) * int(co2, 2))

File: hs/2020/day16/part2.hs
import AOC
import Data.List

count :: Eq a => a -> [a] -> Int
count x = length . filter (==x)

findFields :: [[Bool]] -> [(Int, Int)] -> [(Int, Int)]
findFields [] c = c
findFields t c = case elemIndex 1 ((map (count True)) (transpose t)) of 
    Just n -> case elemIndex True ((transpose t) !! n) of 
        Just k -> findFields ((take k t) ++ [[False | x <- [1..(length (head t))]]] ++ (drop (k+1) t)) ((k, n):c)
    Nothing -> c

main = do
    inp <- readInput
    let [ranges, yourTicket, nearbyTickets] = splitOn "\n\n" inp
    let range = map (nub . concat) (map (\x -> let y = (reverse . words) x in map ((\k -> [(k !! 0)..(k !! 1)]) . (map read) . (splitOn "-")) [y !! 0, y !! 2]) (lines ranges)) :: [[Int]]
    let urange = (nub . concat) range
    let tickets = ((map (map read)) . (map (splitOn ",")) . lines . (drop 16)) nearbyTickets :: [[Int]]
    let filteredTickets = filter (all (flip elem urange)) tickets
    let myTicket = map read (((splitOn ",") . (drop 13)) yourTicket) :: [Int]
    let t = map (\y -> (map (\x -> all (flip elem (range !! y)) ((transpose filteredTickets) !! x))) [0..(length range - 1)]) [0..(length range - 1)]
    let a = findFields t []
    print $ product (map (myTicket !!) (map snd (take 6 (sort a))))
